src/txn/txn_ckpt.c

```diff
@@ -189,10 +211,19 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
 			continue;
 		WT_WITH_DHANDLE(session, session->ckpt_handle[i],
 		    ret = (*op)(session, cfg));
-		WT_RET(ret);
+		WT_ERR(ret);
 	}
 
-	return (0);
+err:
+	/*
+	 * If we have an error somewhere in processing the handles, then
+	 * we need to mark earlier trees dirty.  Recursively call the apply
+	 * function to reset on error.
+	 */
+	if (ret != 0)
+		(void)__checkpoint_apply(session, cfg, __checkpoint_err_reset);
```

Comment:
I'm with @agorrod in not liking the recursion, and with @sueloverso in preferring the error handling here. I don't see any reason to use recursion, let's just duplicate the loop:
```
diff --git a/src/txn/txn_ckpt.c b/src/txn/txn_ckpt.c
index 71f5931d3..d881f3ed8 100644
--- a/src/txn/txn_ckpt.c
+++ b/src/txn/txn_ckpt.c
@@ -173,28 +173,6 @@ err:       __wt_scr_free(session, &tmp);
 }
 
 /*
- * __checkpoint_err_reset --
- *     Reset the modified flag of a btree.  This tree may have been
- *     flushed earlier in the list of all handles and a later tree
- *     got an error.  But metadata is still outstanding.  So reset
- *     on error if needed.
- */
-static int
-__checkpoint_err_reset(WT_SESSION_IMPL *session, const char *cfg[])
-{
-       WT_BTREE *btree;
-
-       WT_UNUSED(cfg);
-       btree = S2BT(session);
-
-       if (btree->was_modified)
-               btree->modified = true;
-       btree->was_modified = false;
-
-       return (0);
-}
-
-/*
  * __checkpoint_apply --
  *     Apply an operation to all handles locked for a checkpoint.
  */
@@ -202,6 +180,7 @@ static int
 __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
        int (*op)(WT_SESSION_IMPL *, const char *[]))
 {
+       WT_BTREE *btree;
        WT_DECL_RET;
        u_int i;
 
@@ -213,16 +192,21 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
                    ret = (*op)(session, cfg));
                WT_ERR(ret);
        }
+       return (0);
 
-err:
        /*
-        * If we have an error somewhere in processing the handles, then
-        * we need to mark earlier trees dirty.  Recursively call the apply
-        * function to reset on error.
+        * If we have an error somewhere in processing the handles, set trees
+        * that were marked clean back to dirty: their associated metadata
+        * changes won't be written and future checkpoints cannot skip them.
         */
-       if (ret != 0)
-               (void)__checkpoint_apply(session, cfg, __checkpoint_err_reset);
-
+err:   for (i = 0; i < session->ckpt_handle_next; ++i) {
+               if (session->ckpt_handle[i] == NULL)
+                       continue;
+               btree = S2BT(session);
+               if (btree->was_modified)
+                       btree->modified = true;
+               btree->was_modified = false;
+       }
        return (ret);
 }
```